File: app/src/main/java/com/agrogames/islandsofwar/islands/impl/IslandFactory.java
package com.agrogames.islandsofwar.islands.impl;

import android.content.Context;

import com.agrogames.islandsofwar.engine.abs.unit.IUnit;
import com.agrogames.islandsofwar.factories.UnitFactory;
import com.agrogames.islandsofwar.islands.abs.Island;
import com.agrogames.islandsofwar.map.impl.Map;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class IslandFactory {
    public static Island parse(Context context, int id, String text) throws JSONException {
        JSONObject json = new JSONObject(text);
        Map map = Map.fromAssets(context, "map" + json.getInt("map_id") + ".txt");
        JSONArray unitsJson = json.getJSONArray("units");
        IUnit[] units = new IUnit[unitsJson.length()];
        for (int i = 0; i < unitsJson.length(); i++){
            JSONObject unit = unitsJson.getJSONObject(i);
            if(unit.has("content")){
                JSONArray contentJson = unit.getJSONArray("content");
                String[] content = new String[contentJson.length()];
                for (int j = 0; j < contentJson.length(); j++) {
                    content[j] = contentJson.getString(j);
                }
                units[i]  = UnitFactory.get(
                        unit.getString("name"),
                        unit.getInt("x"),
                        unit.getInt("y"),
                        content);
            } else {
                units[i]  = UnitFactory.get(
                        unit.getString("name"),
                        unit.getInt("x"),
                        unit.getInt("y"));
            }
            if(unit.has("health")){
                units[i].loseHealth(units[i].getHealth().start - unit.getInt("health"));
            }
        }
        JSONObject a = json.getJSONObject("ship");
        JSONArray contentJson = a.getJSONArray("content");
        String[] content = new String[contentJson.length()];
        for (int j = 0; j < contentJson.length(); j++) {
            content[j] = contentJson.getString(j);
        }
        IUnit attacker  = UnitFactory.get(
                a.getString("name"),
                1,
                1,
                content);
        return new Island(id, map, units, attacker);
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/islands/impl/IslandFactory.kt
package com.agrogames.islandsofwar.islands.impl

import android.content.Context
import com.agrogames.islandsofwar.factories.UnitFactory
import com.agrogames.islandsofwar.islands.abs.Island
import com.agrogames.islandsofwar.map.impl.Map
import org.json.JSONObject

object IslandFactory {
    fun parse(context: Context, id: Int, text: String): Island {
        val json = JSONObject(text)
        val map: Map = Map.fromJsonBin("map" + json.getInt("map_id") + ".txt")
        val unitsJson = json.getJSONArray("units")
        val units = (0 until unitsJson.length()).map {
            val unit = unitsJson.getJSONObject(it)
            val result =  if (unit.has("content")) {
                val contentJson = unit.getJSONArray("content")
                val content = (0 until contentJson.length()).map { contentJson.getString(it) }
                UnitFactory.get(
                        unit.getString("name"),
                        unit.getInt("x"),
                        unit.getInt("y"),
                        *content.toTypedArray())
            } else {
                UnitFactory.get(
                        unit.getString("name"),
                        unit.getInt("x"),
                        unit.getInt("y"))
            }
            if (unit.has("health")) {
                result.loseHealth(result.health.start - unit.getInt("health"))
            }
            result;
        }
        val a = json.getJSONObject("ship")
        val contentJson = a.getJSONArray("content")
        val content = (0 until contentJson.length()).map { contentJson.getString(it) }.toTypedArray()
        val attacker = UnitFactory.get(
                a.getString("name"),
                1,
                1,
                *content)
        return Island(id, map, units.toTypedArray(), attacker)
    }
}